File: java-provider-spark/src/main/java/com/hunorkovacs/koauth/sample/java/providerspark/OauthServlet.java
package com.hunorkovacs.koauth.sample.java.providerspark;

import com.hunorkovacs.koauth.domain.*;
import com.hunorkovacs.koauthsync.domain.mapper.ResponseMapper;
import com.hunorkovacs.koauthsync.domain.mapper.RequestMapper;
import com.hunorkovacs.koauthsync.service.provider.ProviderService;
import com.hunorkovacs.koauthsync.service.provider.ProviderServiceFactory;
import com.hunorkovacs.koauth.service.DefaultTokenGenerator;

import scala.Option;
import scala.concurrent.ExecutionContext;
import scala.util.Either;
import spark.Request;
import spark.Response;
import spark.Route;

import javax.servlet.http.HttpServletResponse;

import static spark.Spark.*;

public class OauthServlet {

    public static final String CONSUMER_KEY = "OmFjJKNqU4v791CWj6QKaBaiEep0WBxJ";

    private ExecutionContext ec = ExecutionContext.Implicits$.MODULE$.global();
    private MyExampleMemoryPersistence persistence = new MyExampleMemoryPersistence(ec);
    private ProviderService provider = ProviderServiceFactory.createProviderService(persistence, ec);
    private RequestMapper<Request> requestMapper = new SparkRequestMapper();

    private class RequestTokenRoute implements Route {
        public Object handle(Request request, spark.Response response) throws Exception {
            KoauthRequest koauthRequest = requestMapper.map(request);
            KoauthResponse koauthResponse = provider.requestToken(koauthRequest);
            return new SparkResponseMapper(response).map(koauthResponse);
        }
    }

    private class AuthorizeRoute implements Route {
        public Object handle(Request request, spark.Response response) throws Exception {
            String username = request.params("username");
            if (username.equals("admin") && request.params("password").equals("admin")) {
                String requestToken = request.params("requestToken");
                String verifier = DefaultTokenGenerator.generateVerifier();
                persistence.authorizeRequestToken(requestToken, username, verifier);
                Option<String> callbackUrl = persistence.getCallback(CONSUMER_KEY, requestToken);
                response.redirect(callbackUrl.get() + "?" + OauthParams.VerifierName() + "=" + verifier);
            } else {
                response.status(401);
                response.body("Credentials or token not valid.");
            }
            return response;
        }
    }

    private class AccessTokenRoute implements Route {
        public Object handle(Request request, spark.Response response) throws Exception {
            KoauthRequest koauthRequest = requestMapper.map(request);
            KoauthResponse koauthResponse = provider.accessToken(koauthRequest);
            return new SparkResponseMapper(response).map(koauthResponse);
        }
    }

    private class MeRoute implements Route {
        public Object handle(Request request, Response response) throws Exception {
            KoauthRequest koauthRequest = requestMapper.map(request);
            Either<KoauthResponse, String> authentication = provider.oauthenticate(koauthRequest);
            if (authentication.isLeft()) {
                KoauthResponse left = authentication.left().get();
                if (left.getClass().equals(ResponseUnauthorized.class)) {
                    response.status(401);
                    return "You are treated as a guest.\n" + ((ResponseUnauthorized) left).body();
                } else {
                    response.status(400);
                    return "You are treated as a guest.\n" + ((ResponseBadRequest) left).body();
                }
            } else {
                String username = authentication.right().get();
                return "You are " + username + ".";
            }
        }
    }

    public static void main(String[] args) {
        OauthServlet twitter = new OauthServlet();
        post("/oauth/request-token", twitter.new RequestTokenRoute());
        get("/oauth/authorize-with-password", twitter.new AuthorizeRoute());
        post("/oauth/access-token", twitter.new AccessTokenRoute());
        get("/me", twitter.new MeRoute());
    }

    private class SparkRequestMapper implements RequestMapper<Request> {
        public KoauthRequest map(Request source) {
            Option<String> body;
            if (source.contentLength() > 0) {
                body = Option.apply(source.body());
            } else {
                body = Option.empty();
            }
            return KoauthRequest.apply(source.requestMethod(), source.url(),
                    Option.apply(source.headers("Authorization")), body);
        }
    }

    private class SparkResponseMapper implements ResponseMapper<spark.Response> {

        private Response response;

        private SparkResponseMapper(Response response) {
            this.response = response;
        }

        public spark.Response map(KoauthResponse source) {
            if (source.getClass().equals(ResponseOk.class)) {
                response.body(((ResponseOk) source).body());
            } else {
                if (source.getClass().equals(ResponseUnauthorized.class)) {
                    response.status(401);
                    response.body(((ResponseUnauthorized) source).body());
                } else {
                    response.status(400);
                    response.body(((ResponseBadRequest) source).body());
                }
            }
            return response;
        }
    }
}

File: scala-provider-scalatra/src/main/scala/com/hunorkovacs/koauth/sample/scala/providerscalatra/OauthServlet.scala
package com.hunorkovacs.koauth.sample.scala.providerscalatra

import javax.servlet.http.HttpServletRequest
import com.hunorkovacs.koauth.domain._
import com.hunorkovacs.koauth.domain.mapper.{ResponseMapper, RequestMapper}
import com.hunorkovacs.koauth.service.DefaultTokenGenerator.generateVerifier
import com.hunorkovacs.koauth.service.provider.ProviderServiceFactory
import com.hunorkovacs.koauth.service.provider.persistence.{RequestToken, ExampleMemoryPersistence}
import org.scalatra._

import scala.concurrent.duration._

import scala.concurrent.{Await, Future, ExecutionContext}

class OauthServlet extends ScalatraServlet {

  implicit private val ec = ExecutionContext.Implicits.global
  private val persistence = new MyExampleMemoryPersistence(ec)
  private val provider = ProviderServiceFactory.createProviderService(persistence, ec)
  private val requestMapper = new JavaxRequestMapper(ec)
  private val responseMapper = new JavaxResponseMapper(ec)

  post("/oauth/request-token") {
    Await.result(mapCallMap(provider.requestToken), 4 seconds)
  }

  private def mapCallMap(f: KoauthRequest => Future[KoauthResponse]) = {
    requestMapper.map(request)
      .flatMap(f)
      .flatMap(responseMapper.map)
  }

  get("/oauth/authorize-with-password") {
    val username = params("username")
    if (username == "admin" && params("password") == "admin") {
      val requestToken = params("requestToken")
      val verifier = generateVerifier
      val callbackUrlF =
        persistence.authorizeRequestToken(requestToken, username, verifier) flatMap { _ =>
          persistence.getCallback(MyExampleMemoryPersistence.ConsumerKey, requestToken)
        }
      val callbackUrl = Await.result(callbackUrlF, 4 seconds)
      redirect(callbackUrl.get + "?" + OauthParams.VerifierName + "=" + verifier)
    } else
      Unauthorized("Credentials or token not valid.")
  }

  post("/oauth/access-token") {
    Await.result(mapCallMap(provider.accessToken), 4 seconds)
  }

  /**
   * But usually these endpoints that serve the main application logic shouldn't be designed exactly like this.
   * In order to save time and code duplication, and focus responsibilities, there should be a separate application
   * whose job is only the authentication verification that proxies all the normal endpoints. It does the first part
   * of the function below: provider.oauthenticate(). Forward incoming requests, complementing them with either the
   * username of the authenticated requester or mark that it couldn't be authenticated and should be handled as
   * a guest. These complemented requests, once gone through the authentication phase, hit the main application,
   * directly with the info of who is making them.
   */
  get("/me") {
    val response =
      requestMapper.map(request) flatMap { koauthRequest =>
        provider.oauthenticate(koauthRequest)
      } map {
        case Left(result) => result match {
          case ResponseUnauthorized(body) => Unauthorized("You are treated as a guest.\n" + body)
          case ResponseBadRequest(body) => BadRequest("You are treated as a guest.\n" + body)
        }
        case Right(username) => Ok("You are " + username + ".")
      }
    Await.result(response, 4 seconds)
  }
}

class JavaxRequestMapper(private val ec: ExecutionContext) extends RequestMapper[HttpServletRequest] {

  implicit private val implicitEc = ec

  override def map(source: HttpServletRequest) = {
    Future {
      val length = source.getContentLength
      val body =
        if (length > 0) {
          val body = new Array[Char](length)
          source.getReader.read(body)
          Some(new String(body))
        } else None
      KoauthRequest(source.getMethod, source.getRequestURL.toString, Option(source.getHeader("Authorization")), body)
    }
  }
}

class JavaxResponseMapper(private val ec: ExecutionContext) extends ResponseMapper[ActionResult] {

  implicit private val implicitEc = ec

  override def map(source: KoauthResponse): Future[ActionResult] = {
    Future {
      source match {
        case ResponseOk(body) => Ok(body)
        case ResponseUnauthorized(body) => Unauthorized(body)
        case ResponseBadRequest(body) => BadRequest(body)
        case _ => InternalServerError()
      }
    }
  }
}

class MyExampleMemoryPersistence(ec: ExecutionContext) extends ExampleMemoryPersistence(ec) {

  implicit private val implicitEc = ec

  def authorizeRequestToken(requestToken: String, verifierUsername: String, verifier: String) = {
    Future {
      requestTokens.find(t => t.requestToken == requestToken) match {
        case None => Unit
        case Some(token) =>
          requestTokens.remove(requestTokens.indexOf(token))
          requestTokens += RequestToken(MyExampleMemoryPersistence.ConsumerKey, token.requestToken,
            token.requestTokenSecret, "", Some(verifierUsername), Some(verifier))
          Unit
      }
    }
  }
}

object MyExampleMemoryPersistence {
  val ConsumerKey = "OmFjJKNqU4v791CWj6QKaBaiEep0WBxJ"
}